File: 235/python/solutions.py
from nodes import BinaryTreeNode


class Solution1:
    # Recursion
    def lowestCommonAncestor_1(self, root: BinaryTreeNode, p: BinaryTreeNode, q: BinaryTreeNode) -> BinaryTreeNode:
        if p.val < root.val > q.val:
            return self.lowestCommonAncestor(root.left, p, q)
        if p.val > root.val < q.val:
            return self.lowestCommonAncestor(root.right, p, q)
        return root

    # while loop
    def lowestCommonAncestor(self, root: BinaryTreeNode, p: BinaryTreeNode, q: BinaryTreeNode) -> BinaryTreeNode:
        while root:
            if p.val < root.val > q.val:
                root = root.left
            elif p.val > root.val < q.val:
                root = root.right
            else:
                return root

File: 235/cpp/solutions.cpp
#include "nodes.h"
#include "lib/trees.h"

using namespace dsa::nodes;
using namespace dsa::lib;

class Solution1
{
public:
    BinaryTreeNode *lowestCommonAncestor(BinarySearchTree root, BinaryTreeNode *p, BinaryTreeNode *q)
    {
        BinarySearchTree *bst;
        bst = new BinarySearchTree(root);
        return bst->lowestCommonAncestor(p, q);
    }
};